In [1]:
!pip install -r requirements.txt

Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 35.3/35.3 MB 48.8 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.2/11.2 MB 86.2 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 285.1/285.1 KB 32.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 78.8 MB/s eta 0:00:00ta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 255.9/255.9 MB 9.7 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 200.4/200.4 KB 25.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 648.0/648.0 KB 50.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.7/233.7 KB 31.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6

In [2]:
!pip install --index-url https://pypi.artifacts.furycloud.io \
    replib.descriptors.tfhub

Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
Looking in indexes: https://pypi.artifacts.furycloud.io
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 394.5/394.5 MB 4.6 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 108.8/108.8 KB 15.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 79.7 MB/s eta 0:00:00:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.9/5.9 MB 77.0 MB/s eta 0:00:00:00:0100:01
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 KB 11.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 462.0/462.0 KB 41.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.0/132.0 KB 22.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 KB 9.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.

In [1]:
import textdistance as td
import json
import base64
import pandas as pd
import csv
import numpy as np
import pyarrow
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV, StratifiedKFold, train_test_split
import matplotlib.pyplot as plt
import pickle
import seaborn as sns
import xgboost
import numpy as np
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from plotnine import *
from melitk.melipass import get_secret
import textacy.preprocessing as tpr
from sklearn.neighbors import NearestNeighbors
from replib.descriptors.tfhub import TensorFlowHub
from tqdm import tqdm
import annoy
import tensorflow_hub as hub
import tensorflow_text

/root/.cache/pypoetry/virtualenvs/app-jhFzdacP-py3.7/lib/python3.7/site-packages/requests/__init__.py:104: RequestsDependencyWarning: urllib3 (1.26.7) or chardet (5.0.0)/charset_normalizer (2.0.10) doesn't match a supported version!
2022-09-29 11:00:59.273490: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-09-29 11:00:59.273534: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2022-09-29 11:01:00.681682: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2022-09-29 11:01:00.681884: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2022-09-29 11:01:00.681

In [3]:
print(hub.__version__)

0.12.0


In [4]:
from google.cloud import bigquery
from google.oauth2 import service_account
class ConnBigQuery:
    def __init__(self, service_account):
        self.service_account = service_account
        self.credentials = self._credentials()
        self.client = self._client()
    def _client(self):
        return bigquery.Client(
            credentials=self.credentials, project=self.credentials.project_id
        )
    def _credentials(self):
        return service_account.Credentials.from_service_account_info(
            self.service_account,
            scopes=["https://www.googleapis.com/auth/cloud-platform"],
        )
    def execute_response(self, query):
        rows = self.client.query(query).result()
        return [dict(row) for row in rows]
    def create_table(self, table_id, schema):
        table = bigquery.Table(table_id, schema=schema)
        self.client.create_table(table)
    def delete_table(self, table_id):
        table = bigquery.Table(table_id)
        self.client.delete_table(table, not_found_ok=True)
    def dataframe_to_table(self, table, df, schema):
        self.client.insert_rows_from_dataframe(
            table=table, dataframe=df, selected_fields=schema
        )

In [5]:
credencialesBQ = eval(base64.b64decode(get_secret("SECRET_BQ")))
bq_connec = ConnBigQuery(credencialesBQ)

In [6]:
model = TensorFlowHub(
    pre_fitted_version="universal-sentence-encoder-multilingual-v3"
)

In [7]:
embed = hub.load("https://tfhub.dev/google/universal-sentence-encoder-multilingual/3")

2022-09-29 11:02:07.169174: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-09-29 11:02:07.169557: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2022-09-29 11:02:09.359249: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2022-09-29 11:02:09.398423: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2499995000 Hz


# 1. Lectura de datos

In [8]:
assortment_df = pd.read_csv('mlb_nielsen_products.csv')
assortment_df = assortment_df[['Categoria', 'SubCategoria', 'Descrição Produto']]
assortment_df.drop_duplicates(inplace=True)

In [9]:
assortment_df.columns = ['category', 'subcategory', 'title']
assortment_df["id"] = assortment_df.index

In [10]:
assortment_df.head(10)

category                 subcategory  \
0  Alimentos e Bebidas                  Chocolates   
1  Alimentos e Bebidas                  Chocolates   
2  Alimentos e Bebidas                  Chocolates   
3  Alimentos e Bebidas                      Whisky   
4  Alimentos e Bebidas                 Leite em Pó   
5  Alimentos e Bebidas                      Whisky   
6  Alimentos e Bebidas                       Vinho   
7  Alimentos e Bebidas                  Salgadinho   
8  Alimentos e Bebidas                 Leite em Pó   
9  Alimentos e Bebidas  Cereal / Proteína em Barra   

                                               title  id  
0          Chocolate Kitkat 4 Fingers Ao Leite 41,5g   0  
1                           CAIXA BOMBOM 250G GAROTO   1  
2            CAIXA BOMBOM ESPECIALIDADES 251G NESTLE   2  
3  Whisky%20Johnnie%20Walker%20Red%20Label%20Esco...   3  
4                         Leite em Pó Aptamil 2 800g   4  
5         Whisky Buchanans Deluxe 12 anos Blended 1L   5  
6          Punta Negra Reserva Tradición Malbec 2019   6  
7                      BATATA PRINGLES ORIGINAL 114G   7  
8                Leite em Pó Danone Aptanutri 3 800g   8  
9  Barra de proteína com recheio - Growth Supplem...   9

In [27]:
des = model.transform(assortment_df.title.iloc[10:11].values)

In [18]:
type(des)

numpy.ndarray

In [28]:
des1 = np.array(embed(assortment_df.title.iloc[10:11].values))

In [29]:
type(des1)

numpy.ndarray

In [30]:
np.array_equal(des, des1)

True

In [21]:
table = 'meli-bi-data.ML_DEV.TMP_ASSORTMENT_MLB_NIELSEN_ITEMS'

schema = [bigquery.SchemaField("category", bigquery.enums.SqlTypeNames.STRING),
          bigquery.SchemaField("subcategory", bigquery.enums.SqlTypeNames.STRING),
          bigquery.SchemaField("title", bigquery.enums.SqlTypeNames.STRING),
          bigquery.SchemaField("id", bigquery.enums.SqlTypeNames.INT64)
         ]

for df_chunk in np.array_split(assortment_df, 10):
    job_config = bigquery.LoadJobConfig(schema=schema)
    job_config.write_disposition = bigquery.WriteDisposition.WRITE_APPEND
    # Load data to BQ
    job = bq_connec.client.load_table_from_dataframe(df_chunk, table, job_config=job_config)
    job.result()

# 2. Preprocesamiento de datos

In [8]:
def title_preprocesing(df):
    df = df.apply(lambda x: tpr.remove.accents(x))
    df = df.apply(lambda x: tpr.replace.urls(x, repl=''))
    df = df.apply(lambda x: tpr.replace.hashtags(x, repl=''))
    df = df.apply(lambda x: tpr.replace.currency_symbols(x, repl=''))
    df = df.apply(lambda x: tpr.remove.punctuation(x))
    df = df.apply(lambda x: x.lower())
    
    return df

In [9]:
assortment_modified_df = assortment_df

In [10]:
assortment_modified_df['category'] = title_preprocesing(assortment_modified_df['category'])
assortment_modified_df['subcategory'] = title_preprocesing(assortment_modified_df['subcategory'])
assortment_modified_df['title'] = title_preprocesing(assortment_modified_df['title'])

assortment_modified_df['cat_complete'] = assortment_modified_df.apply(lambda x: x.category + ' ' + x.subcategory, axis = 1)
assortment_modified_df = assortment_modified_df.reset_index(drop = True)

assortment_modified_df.head(10)

category                 subcategory  \
0  alimentos e bebidas                  chocolates   
1  alimentos e bebidas                  chocolates   
2  alimentos e bebidas                  chocolates   
3  alimentos e bebidas                      whisky   
4  alimentos e bebidas                 leite em po   
5  alimentos e bebidas                      whisky   
6  alimentos e bebidas                       vinho   
7  alimentos e bebidas                  salgadinho   
8  alimentos e bebidas                 leite em po   
9  alimentos e bebidas  cereal   proteina em barra   

                                               title  id  \
0          chocolate kitkat 4 fingers ao leite 41 5g   0   
1                           caixa bombom 250g garoto   1   
2            caixa bombom especialidades 251g nestle   2   
3  whisky 20johnnie 20walker 20red 20label 20esco...   3   
4                         leite em po aptamil 2 800g   4   
5         whisky buchanans deluxe 12 anos blended 1l   5   
6          punta negra reserva tradicion malbec 2019   6   
7                      batata pringles original 114g   7   
8                leite em po danone aptanutri 3 800g   8   
9  barra de proteina com recheio   growth supplem...   9   

                                     cat_complete  
0                  alimentos e bebidas chocolates  
1                  alimentos e bebidas chocolates  
2                  alimentos e bebidas chocolates  
3                      alimentos e bebidas whisky  
4                 alimentos e bebidas leite em po  
5                      alimentos e bebidas whisky  
6                       alimentos e bebidas vinho  
7                  alimentos e bebidas salgadinho  
8                 alimentos e bebidas leite em po  
9  alimentos e bebidas cereal   proteina em barra

# 3. Cálculo de Embeddings

In [16]:
model = TensorFlowHub(
    pre_fitted_version="universal-sentence-encoder-multilingual-v3"
)

In [11]:
chunk_size = 10000
chunks = int(np.ceil(assortment_modified_df.shape[0] / chunk_size))

print(chunks)

1


In [12]:
import time

emb_array = np.zeros((1,512))

for i in range(chunks):
    t0 = time.time()
    emb_array = np.append(
        emb_array, 
        model.transform(assortment_modified_df.cat_complete.iloc[i*chunk_size:(i+1)*chunk_size].tolist()),
        axis = 0
    )
    t1 = time.time()

    print(f'Chunk #{i}: {round(t1-t0, 2)} seconds')

2022-08-30 12:29:27.802372: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-08-30 12:29:27.802711: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2022-08-30 12:29:32.116141: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2022-08-30 12:29:32.158701: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2499995000 Hz


/root/.cache/pypoetry/virtualenvs/app-jhFzdacP-py3.7/lib/python3.7/site-packages/sklearn/base.py:338: UserWarning: Trying to unpickle estimator Pipeline from version 0.24.1 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations


Chunk #0: 9.11 seconds


In [13]:
emb_array = np.delete(emb_array, 0, 0)
emb_array.shape

(6414, 512)

# 4. Categorias

## 4.1 Categorias (CAT+SUBCAT)

El primer paso es hacer similaridad entre categoría + subcategoría de nielsen contra L1 + L3 de MeLi

In [14]:
query = """
    SELECT DISTINCT
        SIT_SITE_ID,

        CAT_CATEG_ID_L1,
        CAT_CATEG_ID_L2,
        CAT_CATEG_ID_L3,

        CAT_CATEG_NAME_L1,
        CAT_CATEG_NAME_L2,
        CAT_CATEG_NAME_L3,

        CAT_DOMAIN_ID

    FROM
        `meli-bi-data.WHOWNER.AG_LK_CAT_CATEGORIES_PH`
    WHERE
        SIT_SITE_ID = 'MLB'
    AND
        PHOTO_ID = 'TODATE'
    AND
        CAT_DOMAIN_ID IS NOT NULL
"""

In [15]:
categories_df = pd.DataFrame(bq_connec.execute_response(query))

In [16]:
categories_df.shape

(10499, 8)

In [17]:
categories_df['CAT_CATEG_NAME_L1'] = title_preprocesing(categories_df['CAT_CATEG_NAME_L1'])
categories_df['CAT_CATEG_NAME_L3'] = title_preprocesing(categories_df['CAT_CATEG_NAME_L3'])

categories_df['CAT_COMPLETE'] = categories_df.apply(lambda x: x.CAT_CATEG_NAME_L1 + ' ' + x.CAT_CATEG_NAME_L3, axis = 1)
categories_df = categories_df.reset_index(drop = True)

categories_df.head(10)

SIT_SITE_ID CAT_CATEG_ID_L1 CAT_CATEG_ID_L2 CAT_CATEG_ID_L3  \
0         MLB          218519          225897          225897   
1         MLB            1459            1466            1468   
2         MLB            1459          214910          214916   
3         MLB            1540            9141            9138   
4         MLB          218519          255707          255707   
5         MLB            1196          455868          455868   
6         MLB            1540           63955           64123   
7         MLB            1459          214910          214914   
8         MLB            1459          215108          215112   
9         MLB            1459            1472            1479   

           CAT_CATEG_NAME_L1     CAT_CATEG_NAME_L2  \
0                  ingressos  Ingressos de Coleção   
1                    imoveis                 Casas   
2                    imoveis                Sítios   
3                   servicos      Festas e Eventos   
4                  ingressos                 Shows   
5  livros  revistas e comics                Ebooks   
6                   servicos               Animais   
7                    imoveis                Sítios   
8                    imoveis               Galpões   
9                    imoveis          Apartamentos   

              CAT_CATEG_NAME_L3                        CAT_DOMAIN_ID  \
0  level 3 ingressos de colecao                    MLB-EVENT_TICKETS   
1                         venda      MLB-DEVELOPMENT_HOUSES_FOR_SALE   
2                     temporada  MLB-FARM_HOUSES_FOR_VACATION_RENTAL   
3                       equipes                   MLB-OTHER_SERVICES   
4                 level 3 shows                  MLB-CONCERT_TICKETS   
5                level 3 ebooks                           MLB-EBOOKS   
6                 hotel e canil                     MLB-PET_SERVICES   
7                         venda             MLB-FARM_HOUSES_FOR_SALE   
8                       aluguel              MLB-WAREHOUSES_FOR_RENT   
9                       aluguel              MLB-APARTMENTS_FOR_RENT   

                               CAT_COMPLETE  
0    ingressos level 3 ingressos de colecao  
1                             imoveis venda  
2                         imoveis temporada  
3                          servicos equipes  
4                   ingressos level 3 shows  
5  livros  revistas e comics level 3 ebooks  
6                    servicos hotel e canil  
7                             imoveis venda  
8                           imoveis aluguel  
9                           imoveis aluguel

In [18]:
chunk_size = 10000
chunks = int(np.ceil(categories_df.shape[0] / chunk_size))

print(chunks)

2


In [19]:
import time

emb_cat_array = np.zeros((1,512))

for i in range(chunks):
    t0 = time.time()
    emb_cat_array = np.append(
        emb_cat_array, 
        model.transform(categories_df.CAT_COMPLETE.iloc[i*chunk_size:(i+1)*chunk_size].tolist()),
        axis = 0
    )
    t1 = time.time()

    print(f'Chunk #{i}: {round(t1-t0, 2)} seconds')

Chunk #0: 3.3 seconds
Chunk #1: 0.89 seconds


In [20]:
emb_cat_array = np.delete(emb_cat_array, 0, 0)
emb_cat_array.shape

(10499, 512)

In [21]:
!pip install annoy

Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
You should consider upgrading via the '/usr/local/bin/python3 -m pip install --upgrade pip' command.


In [23]:
n, p = emb_cat_array.shape
n_trees_from_data = int(np.sqrt(p))
n_trees_from_data

22

In [24]:
t = annoy.AnnoyIndex(p, "euclidean")
for idx, row in enumerate(emb_cat_array):
    t.add_item(idx, row)
    
t.build(n_trees_from_data)

True

In [25]:
columns_names = ['id','category','subcategory','title','CAT_DOMAIN_ID']
complete_df = pd.DataFrame(columns=columns_names)

for item in range(emb_array.shape[0]):
    idx, dist = t.get_nns_by_vector(emb_array[item], n=25, include_distances=True)

    neighbours_df = pd.DataFrame(columns=columns_names)

    for i, element in enumerate(dist):
        if element < 0.6:
            aux_df = pd.DataFrame(pd.concat(
                [
                    assortment_df.iloc[item][['id','category','subcategory','title']],
                    categories_df.iloc[idx[i]][['CAT_DOMAIN_ID']]
                ]
            )
            ).T

            neighbours_df = pd.concat([aux_df[columns_names], neighbours_df], ignore_index=True)
            
    complete_df = pd.concat([neighbours_df, complete_df], ignore_index=True)

In [26]:
complete_df.drop_duplicates(inplace=True)
complete_df.tail(5)

id             category subcategory  \
15517  0  alimentos e bebidas  chocolates   
15518  0  alimentos e bebidas  chocolates   
15519  0  alimentos e bebidas  chocolates   
15520  0  alimentos e bebidas  chocolates   
15521  0  alimentos e bebidas  chocolates   

                                           title         CAT_DOMAIN_ID  
15517  chocolate kitkat 4 fingers ao leite 41 5g        MLB-CHOCOLATES  
15518  chocolate kitkat 4 fingers ao leite 41 5g  MLB-FOODS_AND_DRINKS  
15519  chocolate kitkat 4 fingers ao leite 41 5g       MLB-EASTER_EGGS  
15520  chocolate kitkat 4 fingers ao leite 41 5g           MLB-CANDIES  
15521  chocolate kitkat 4 fingers ao leite 41 5g           MLB-NOUGATS

In [27]:
complete_df.id.drop_duplicates().shape

(2224,)

In [28]:
complete_df[pd.isna(complete_df.CAT_DOMAIN_ID)]

Empty DataFrame
Columns: [id, category, subcategory, title, CAT_DOMAIN_ID]
Index: []

Como se puede ver, con este enfoque y un threshold de 0.6, se encuentra match de categoría y dominio para 2224 items.

## 4.2 Categorias (Title)

El segundo enfoque consiste en usar una muestra de títulos de cada dominio en meli (al menos 500 por dominio) para la similaridad

In [29]:
query = """
    SELECT 
        * 
    FROM 
        `meli-bi-data.ML_DEV.TMP_SHOPEE_DOMAIN_IDS`
"""

In [30]:
domains_df = pd.DataFrame(bq_connec.execute_response(query))

In [31]:
def title_brand(title, brands):
    for item in brands.split(' '):
        brand_add = ''
        if item not in title:
            brand_add += item
            
    return title + ' ' + brand_add

In [32]:
domains_df['BRAND_NAME'] = domains_df['BRAND_NAME'].apply(lambda x: '' if x is None else x)
domains_df['BRAND_NAME'] = title_preprocesing(domains_df['BRAND_NAME'])
domains_df['ITE_ITEM_TITLE'] = title_preprocesing(domains_df['ITE_ITEM_TITLE'])

domains_df['TITLE_BRAND'] = domains_df.apply(lambda x: title_brand(x.ITE_ITEM_TITLE, x.BRAND_NAME), axis = 1)

domains_df = domains_df.reset_index(drop = True)

In [33]:
domains_df.tail(5)

SIT_SITE_ID                ITE_ITEM_DOM_DOMAIN_ID  ITE_ITEM_ID  \
1959793         MLB  MLB-MUSICAL_KEYBOARD_RUBBER_CONTACTS   1943104517   
1959794         MLB  MLB-MUSICAL_KEYBOARD_RUBBER_CONTACTS   1943094931   
1959795         MLB  MLB-MUSICAL_KEYBOARD_RUBBER_CONTACTS   1943100069   
1959796         MLB  MLB-MUSICAL_KEYBOARD_RUBBER_CONTACTS   1931977077   
1959797         MLB                    MLB-TOY_TOOLS_SETS   2644396966   

                                            ITE_ITEM_TITLE  \
1959793  placa p  teclado yamaha psr e353 regua menor m...   
1959794  placa p  teclado yamaha psr e323 regua menor m...   
1959795  placa p  teclado yamaha psre423 regua menor mk...   
1959796  2 placas p  teclado yamaha s750 reguas xr564 +...   
1959797  kit ferramentas infantil engenheiro com capace...   

                                                BRAND_NAME  \
1959793  yamaha psr 290 psr 292 psr 295 psr e 303 psr e...   
1959794  yamaha psr 290 psr 292 psr 295 psr e 303 psr e...   
1959795  yamaha psr 290 psr 292 psr 295 psr e 303 psr e...   
1959796  yamaha psr b50 psr 230 psr 330 psr 520  os ult...   
1959797  kit de ferramentas infantil  kit de ferramenta...   

                                               TITLE_BRAND  
1959793  placa p  teclado yamaha psr e353 regua menor m...  
1959794  placa p  teclado yamaha psr e323 regua menor m...  
1959795  placa p  teclado yamaha psre423 regua menor mk...  
1959796  2 placas p  teclado yamaha s750 reguas xr564 +...  
1959797  kit ferramentas infantil engenheiro com capace...

In [34]:
chunk_size = 10000
chunks = int(np.ceil(domains_df.shape[0] / chunk_size))

print(chunks)

196


In [35]:
import time

emb_dom_array = np.zeros((1,512))

for i in range(chunks):
    t0 = time.time()
    emb_dom_array = np.append(
        emb_dom_array, 
        model.transform(domains_df.TITLE_BRAND.iloc[i*chunk_size:(i+1)*chunk_size].tolist()),
        axis = 0
    )
    t1 = time.time()

    print(f'Chunk #{i}: {round(t1-t0, 2)} seconds')

Chunk #0: 3.74 seconds
Chunk #1: 3.93 seconds
Chunk #2: 3.96 seconds
Chunk #3: 4.21 seconds
Chunk #4: 4.21 seconds
Chunk #5: 4.17 seconds
Chunk #6: 4.6 seconds
Chunk #7: 4.5 seconds
Chunk #8: 4.51 seconds
Chunk #9: 4.6 seconds
Chunk #10: 4.51 seconds
Chunk #11: 4.7 seconds
Chunk #12: 4.63 seconds
Chunk #13: 4.62 seconds
Chunk #14: 4.42 seconds
Chunk #15: 4.51 seconds
Chunk #16: 4.62 seconds
Chunk #17: 4.69 seconds
Chunk #18: 4.63 seconds
Chunk #19: 4.68 seconds
Chunk #20: 4.81 seconds
Chunk #21: 4.7 seconds
Chunk #22: 4.77 seconds
Chunk #23: 4.71 seconds
Chunk #24: 4.74 seconds
Chunk #25: 4.86 seconds
Chunk #26: 4.86 seconds
Chunk #27: 4.86 seconds
Chunk #28: 4.79 seconds
Chunk #29: 5.07 seconds
Chunk #30: 4.79 seconds
Chunk #31: 4.85 seconds
Chunk #32: 4.88 seconds
Chunk #33: 4.42 seconds
Chunk #34: 4.54 seconds
Chunk #35: 4.55 seconds
Chunk #36: 4.64 seconds
Chunk #37: 4.69 seconds
Chunk #38: 4.68 seconds
Chunk #39: 5.14 seconds
Chunk #40: 5.09 seconds
Chunk #41: 5.06 seconds
Chunk #

In [36]:
emb_dom_array = np.delete(emb_dom_array, 0, 0)
emb_dom_array.shape

(1959798, 512)

In [37]:
n, p = emb_dom_array.shape
n_trees_from_data = int(np.sqrt(p))
n_trees_from_data

22

In [38]:
dom_tree = annoy.AnnoyIndex(p, "euclidean")
for idx, row in enumerate(emb_dom_array):
    dom_tree.add_item(idx, row)
    
dom_tree.build(n_trees_from_data)

True

In [39]:
dom_tree.save('models/domain.tree')

True

Se usa un threshold de 1.0 para la similaridad (a diferencia del 0.6 que se usó con las categorías)

In [40]:
columns_names = ['id','category','subcategory','title','ITE_ITEM_DOM_DOMAIN_ID']
complete_dom_df = pd.DataFrame(columns=columns_names)

for item in range(emb_array.shape[0]):
    idx, dist = dom_tree.get_nns_by_vector(emb_array[item], n=25, include_distances=True)

    neighbours_df = pd.DataFrame(columns=columns_names)

    for i, element in enumerate(dist):
        if element < 1.0:
            aux_df = pd.DataFrame(pd.concat(
                [
                    assortment_df.iloc[item][['id','category','subcategory','title']],
                    domains_df.iloc[idx[i]][['ITE_ITEM_DOM_DOMAIN_ID']]
                ]
            )
            ).T

            neighbours_df = pd.concat([aux_df[columns_names], neighbours_df], ignore_index=True)
    
    neighbours_df.drop_duplicates(inplace=True)
            
    complete_dom_df = pd.concat([neighbours_df, complete_dom_df], ignore_index=True)
    
    if item % 200 == 0:
        print(f'item: {item}')

item: 0
item: 200
item: 400
item: 600
item: 800
item: 1000
item: 1200
item: 1400
item: 1600
item: 1800
item: 2000
item: 2200
item: 2400
item: 2600
item: 2800
item: 3000
item: 3200
item: 3400
item: 3600
item: 3800
item: 4000
item: 4200
item: 4400
item: 4600
item: 4800
item: 5000
item: 5200
item: 5400
item: 5600
item: 5800
item: 6000
item: 6200
item: 6400


In [41]:
complete_dom_df.columns = ['id','category','subcategory','title','CAT_DOMAIN_ID']

complete_dom_df.head(10)

id   category           subcategory  \
0  6413        vhs                   vhs   
1  6413        vhs                   vhs   
2  6413        vhs                   vhs   
3  6413        vhs                   vhs   
4  6412  telefonia  celular e smartphone   
5  6412  telefonia  celular e smartphone   
6  6412  telefonia  celular e smartphone   
7  6412  telefonia  celular e smartphone   
8  6412  telefonia  celular e smartphone   
9  6412  telefonia  celular e smartphone   

                                               title  \
0          apcrifos e pseudo epgrafos da bblia vol 3   
1          apcrifos e pseudo epgrafos da bblia vol 3   
2          apcrifos e pseudo epgrafos da bblia vol 3   
3          apcrifos e pseudo epgrafos da bblia vol 3   
4  smartphone samsung galaxy note20 cinza 256gb  ...   
5  smartphone samsung galaxy note20 cinza 256gb  ...   
6  smartphone samsung galaxy note20 cinza 256gb  ...   
7  smartphone samsung galaxy note20 cinza 256gb  ...   
8  smartphone samsung galaxy note20 cinza 256gb  ...   
9  smartphone samsung galaxy note20 cinza 256gb  ...   

                       CAT_DOMAIN_ID  
0                MLB-VIDEO_CASSETTES  
1                  MLB-EROTIC_MOVIES  
2                  MLB-DRONE_MODULES  
3                         MLB-COMICS  
4                  MLB-TELEPROMPTERS  
5                     MLB-TELEPHONES  
6             MLB-HEADSET_TELEPHONES  
7  MLB-MILITARY_COLLECTIBLE_SUPPLIES  
8                 MLB-CAMERA_TRIPODS  
9             MLB-CELLPHONE_SOFTWARE

In [42]:
complete_dom_df.id.drop_duplicates().shape

(6173,)

In [43]:
assortment_df.shape

(6414, 5)

Se encuentran 6173 items con match de categoría/dominio. La elección del threshold es arbitraria; un valor grande implica que más items tienen match de dominios pero así mismo cada item recibe más dominios lo que aumenta la búsqueda. Al final, los dos resultados son usados.

## 4.3 Concatenation

Se concatenan los resultados anteriores

In [44]:
domain_model_df = pd.concat([complete_df, complete_dom_df], ignore_index=True)
domain_model_df.drop_duplicates(inplace=True)
domain_model_df.head(10)

id   category           subcategory  \
0  6412  telefonia  celular e smartphone   
1  6412  telefonia  celular e smartphone   
2  6412  telefonia  celular e smartphone   
3  6411  telefonia  celular e smartphone   
4  6411  telefonia  celular e smartphone   
5  6411  telefonia  celular e smartphone   
6  6410  telefonia  celular e smartphone   
7  6410  telefonia  celular e smartphone   
8  6410  telefonia  celular e smartphone   
9  6409  telefonia  celular e smartphone   

                                               title      CAT_DOMAIN_ID  
0  smartphone samsung galaxy note20 cinza 256gb  ...     MLB-CELLPHONES  
1  smartphone samsung galaxy note20 cinza 256gb  ...     MLB-TELEPHONES  
2  smartphone samsung galaxy note20 cinza 256gb  ...  MLB-IP_TELEPHONES  
3  smartphone samsung galaxy s21 fe 5g branco 128...     MLB-CELLPHONES  
4  smartphone samsung galaxy s21 fe 5g branco 128...     MLB-TELEPHONES  
5  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-IP_TELEPHONES  
6  iphone 11 apple 128gb branco 6 1 e2 80 9d 12mp...     MLB-CELLPHONES  
7  iphone 11 apple 128gb branco 6 1 e2 80 9d 12mp...     MLB-TELEPHONES  
8  iphone 11 apple 128gb branco 6 1 e2 80 9d 12mp...  MLB-IP_TELEPHONES  
9  iphone 12 apple 128gb ios 5g wi fi tela 6 1   ...     MLB-CELLPHONES

In [45]:
table = 'meli-bi-data.ML_DEV.TMP_ASSORTMENT_MLB_DOMAIN_MATCHES'

schema = [bigquery.SchemaField("id", bigquery.enums.SqlTypeNames.INT64),
          bigquery.SchemaField("category", bigquery.enums.SqlTypeNames.STRING),
          bigquery.SchemaField("subcategory", bigquery.enums.SqlTypeNames.STRING),
          bigquery.SchemaField("title", bigquery.enums.SqlTypeNames.STRING),
          bigquery.SchemaField("CAT_DOMAIN_ID", bigquery.enums.SqlTypeNames.STRING)
         ]

for df_chunk in np.array_split(domain_model_df, 10):
    job_config = bigquery.LoadJobConfig(schema=schema)
    job_config.write_disposition = bigquery.WriteDisposition.WRITE_APPEND
    # Load data to BQ
    job = bq_connec.client.load_table_from_dataframe(df_chunk, table, job_config=job_config)
    job.result()

# 5. Neighbours

In [11]:
query = """
    SELECT
        *
    FROM
        meli-bi-data.ML_DEV.TMP_ASSORTMENT_MLB_DOMAIN_MATCHES
"""

domain_model_df = pd.DataFrame(bq_connec.execute_response(query))

In [13]:
domain_model_df.tail(5)

id                        category                subcategory  \
68160  3777  industria  comercio e negocios  gaveteiro para escritorio   
68161  3777  industria  comercio e negocios  gaveteiro para escritorio   
68162  3777  industria  comercio e negocios  gaveteiro para escritorio   
68163  3777  industria  comercio e negocios  gaveteiro para escritorio   
68164  3777  industria  comercio e negocios  gaveteiro para escritorio   

                                                   title  \
68160  gaveteiro 4 gavetas com rodizios e chave gavea...   
68161  gaveteiro 4 gavetas com rodizios e chave gavea...   
68162  gaveteiro 4 gavetas com rodizios e chave gavea...   
68163  gaveteiro 4 gavetas com rodizios e chave gavea...   
68164  gaveteiro 4 gavetas com rodizios e chave gavea...   

                         CAT_DOMAIN_ID  
68160                MLB-TANDEM_CHAIRS  
68161                   MLB-FILE_BOXES  
68162                MLB-WASTE_BASKETS  
68163  MLB-HOME_OFFICE_DESK_ORGANIZERS  
68164              MLB-OFFICE_CUBICLES

## 5.1 Title Embeddings - NIELSEN 

Test de cosas. Ignorar.

In [17]:
model = TensorFlowHub(
    pre_fitted_version="universal-sentence-encoder-multilingual-v3"
)

In [14]:
chunk_size = 10000
chunks = int(np.ceil(assortment_modified_df.shape[0] / chunk_size))

print(chunks)

1


In [18]:
import time

emb_array = np.zeros((1,512))

for i in range(chunks):
    t0 = time.time()
    emb_array = np.append(
        emb_array, 
        model.transform(assortment_modified_df.title.iloc[i*chunk_size:(i+1)*chunk_size].tolist()),
        axis = 0
    )
    t1 = time.time()

    print(f'Chunk #{i}: {round(t1-t0, 2)} seconds')

2022-08-30 16:08:09.034569: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-08-30 16:08:09.034971: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2022-08-30 16:08:13.231322: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2022-08-30 16:08:13.286689: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2499995000 Hz


/root/.cache/pypoetry/virtualenvs/app-jhFzdacP-py3.7/lib/python3.7/site-packages/sklearn/base.py:338: UserWarning: Trying to unpickle estimator Pipeline from version 0.24.1 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations


Chunk #0: 10.29 seconds


In [19]:
emb_array = np.delete(emb_array, 0, 0)
emb_array.shape

(6414, 512)

## 5.2 Title Embeddings - MELI Items

In [70]:
domains_list = list(domain_model_df.CAT_DOMAIN_ID.drop_duplicates())
print(len(domains_list))
domains_list[:10]

2006


['MLB-CELLPHONES',
 'MLB-TELEPHONES',
 'MLB-IP_TELEPHONES',
 'MLB-WALKIE_TALKIE_CHARGERS',
 'MLB-SMARTWATCH_CHARGERS',
 'MLB-WIRELESS_CHARGERS',
 'MLB-PORTABLE_CELLPHONE_CHARGERS',
 'MLB-CELLPHONE_ACCESSORIES',
 'MLB-MOBILE_DEVICE_CHARGERS',
 'MLB-CELLPHONE_BATTERY_CASES']

In [124]:
cluster_number = 3

query = """
    SELECT 
        SIT_SITE_ID,
        ITE_ITEM_DOM_DOMAIN_ID,
        ITE_ITEM_ID,
        ITE_ITEM_TITLE,
      
    FROM
      `meli-bi-data.ML_TBL.ITEMS_SEGMENTOS`

    WHERE SIT_SITE_ID = 'MLB'
    AND GRUPO IN ('validos', 'incompletos')
    AND ITE_ITEM_DOM_DOMAIN_ID = '{dominios}'
""".format(dominios = domains_list[cluster_number])

meli_items_df = pd.DataFrame(bq_connec.execute_response(query))

meli_items_df['ITE_ITEM_TITLE'] = title_preprocesing(meli_items_df['ITE_ITEM_TITLE'])
meli_items_df = meli_items_df.reset_index(drop = True)

In [125]:
meli_items_df.head(5)

SIT_SITE_ID ITE_ITEM_DOM_DOMAIN_ID  ITE_ITEM_ID  \
0         MLB         MLB-CELLPHONES   1958175911   
1         MLB         MLB-CELLPHONES   2732083726   
2         MLB         MLB-CELLPHONES   2660392032   
3         MLB         MLB-CELLPHONES   2627932413   
4         MLB         MLB-CELLPHONES   2613753601   

                                      ITE_ITEM_TITLE  
0  apple iphone 8 64 gb original pronta entrega v...  
1                celular iphone 11 128gb   excelente  
2                                     celular xiaomi  
3  xiaomi redmi note 11  snapdragon  dual sim 128...  
4  xiaomi note 11 pro dual sim 128 gb preto 8 gb ram

In [126]:
chunk_size = 10000
chunks = int(np.ceil(meli_items_df.shape[0] / chunk_size))

print(chunks)

6


In [127]:
import time

emb_meli_array = np.zeros((1,512))

for i in range(chunks):
    t0 = time.time()
    emb_meli_array = np.append(
        emb_meli_array, 
        model.transform(meli_items_df.ITE_ITEM_TITLE.iloc[i*chunk_size:(i+1)*chunk_size].tolist()),
        axis = 0
    )
    t1 = time.time()

    print(f'Chunk #{i}: {round(t1-t0, 2)} seconds')

Chunk #0: 4.16 seconds
Chunk #1: 4.06 seconds
Chunk #2: 4.08 seconds
Chunk #3: 4.07 seconds
Chunk #4: 4.06 seconds
Chunk #5: 0.88 seconds


In [128]:
emb_meli_array = np.delete(emb_meli_array, 0, 0)
emb_meli_array.shape

(51877, 512)

In [129]:
n, p = emb_meli_array.shape
n_trees_from_data = int(np.sqrt(p))
n_trees_from_data

22

In [130]:
y = annoy.AnnoyIndex(p, "euclidean")
for idx, row in enumerate(emb_meli_array):
    y.add_item(idx, row)
    
y.build(n_trees_from_data)

True

In [172]:
search_df = complete_df[complete_df.CAT_DOMAIN_ID == domains_list[cluster_number]]

columns_names = ['id','category','subcategory','title','CAT_DOMAIN_ID','SIT_SITE_ID','ITE_ITEM_ID','ITE_ITEM_TITLE']
matches_df = pd.DataFrame(columns=columns_names)


for index, row in search_df.iterrows():
    
    idx, dist = y.get_nns_by_vector(emb_array[row.id], n=25, include_distances=True)

    neighbours_df = pd.DataFrame(columns=columns_names)

    for i, element in enumerate(dist):
        if element < 0.75:            
            aux_df = pd.DataFrame(pd.concat(
                [
                    row[['id','category','subcategory','title','CAT_DOMAIN_ID']],
                    meli_items_df.iloc[idx[i]][['SIT_SITE_ID','ITE_ITEM_ID','ITE_ITEM_TITLE']]
                ]
            )
            ).T
            
            neighbours_df = pd.concat([aux_df[columns_names], neighbours_df], ignore_index=True)

    matches_df = pd.concat([neighbours_df, matches_df], ignore_index=True)

In [174]:
matches_df.tail(10)

id   category           subcategory  \
4278  6411  telefonia  celular e smartphone   
4279  6411  telefonia  celular e smartphone   
4280  6411  telefonia  celular e smartphone   
4281  6411  telefonia  celular e smartphone   
4282  6411  telefonia  celular e smartphone   
4283  6411  telefonia  celular e smartphone   
4284  6411  telefonia  celular e smartphone   
4285  6411  telefonia  celular e smartphone   
4286  6411  telefonia  celular e smartphone   
4287  6411  telefonia  celular e smartphone   

                                                  title   CAT_DOMAIN_ID  \
4278  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   
4279  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   
4280  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   
4281  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   
4282  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   
4283  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   
4284  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   
4285  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   
4286  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   
4287  smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES   

     SIT_SITE_ID ITE_ITEM_ID  \
4278         MLB  2640197719   
4279         MLB  2640234916   
4280         MLB  2726526426   
4281         MLB  2726526425   
4282         MLB  2688915859   
4283         MLB  2688929456   
4284         MLB  2644291288   
4285         MLB  2644291291   
4286         MLB  2618590766   
4287         MLB  2184619850   

                                         ITE_ITEM_TITLE  
4278  smartphone samsung galaxy s20 fe 5g 128gb 6gb ...  
4279  smartphone samsung galaxy s20 fe 5g 128gb 6gb ...  
4280  smartphone samsung galaxy s21 plus 256gb + 8gb...  
4281  smartphone samsung galaxy s21 plus 256gb + 8gb...  
4282  smartphone samsung galaxy s20 fe 128gb 5g g781...  
4283  smartphone samsung galaxy s20 fe 128gb 5g g781...  
4284  smartphone samsung galaxy s21 fe 5g preto 128g...  
4285  smartphone samsung galaxy s21 fe 5g preto 128g...  
4286  smartphone samsung galaxy s21 fe 128gb  6gb  6...  
4287  smartphone samsung galaxy s21 fe 5g 128gb  6gb...

In [164]:
matches_df.shape

(4288, 8)

In [150]:
search_df = complete_df[complete_df.CAT_DOMAIN_ID == domains_list[cluster_number]]
search_df

id   category           subcategory  \
4     6412  telefonia  celular e smartphone   
29    6411  telefonia  celular e smartphone   
54    6410  telefonia  celular e smartphone   
79    6409  telefonia  celular e smartphone   
104   6408  telefonia  celular e smartphone   
...    ...        ...                   ...   
4879  6217  telefonia  celular e smartphone   
4904  6216  telefonia  celular e smartphone   
4929  6215  telefonia  celular e smartphone   
4954  6214  telefonia  celular e smartphone   
4979  6213  telefonia  celular e smartphone   

                                                  title   CAT_DOMAIN_ID  
4     smartphone samsung galaxy note20 cinza 256gb  ...  MLB-CELLPHONES  
29    smartphone samsung galaxy s21 fe 5g branco 128...  MLB-CELLPHONES  
54    iphone 11 apple 128gb branco 6 1 e2 80 9d 12mp...  MLB-CELLPHONES  
79    iphone 12 apple 128gb ios 5g wi fi tela 6 1   ...  MLB-CELLPHONES  
104   smartphone motorola moto e7 dual chip android ...  MLB-CELLPHONES  
...                                                 ...             ...  
4879  iphone 11 apple 64gb branco  tela de 6 1 e2 80...  MLB-CELLPHONES  
4904        smartphone samsung galaxy a03 64gb preto 4g  MLB-CELLPHONES  
4929  iphone 11 apple 128gb preto 6 1 e2 80 9d 12mp ios  MLB-CELLPHONES  
4954  iphone 11 apple 64gb preto  tela de 6 1 e2 80 ...  MLB-CELLPHONES  
4979  iphone 13 apple 128gb meia noite tela de 6 1  ...  MLB-CELLPHONES  

[199 rows x 5 columns]

In [151]:
y.get_nns_by_vector(emb_array[6412], n=15, include_distances=True)

([49,
  2651,
  9028,
  45230,
  15687,
  15150,
  45424,
  21074,
  27469,
  36404,
  18518,
  8534,
  22727,
  1964,
  25436],
 [0.7909523844718933,
  0.7909523844718933,
  0.7909523844718933,
  0.7909523844718933,
  0.8059686422348022,
  0.8153699636459351,
  0.8153699636459351,
  0.8161715865135193,
  0.8161715865135193,
  0.8222026228904724,
  0.8260628581047058,
  0.8352891802787781,
  0.8401623368263245,
  0.8405235409736633,
  0.8405235409736633])

In [152]:
meli_items_df.iloc[49]

SIT_SITE_ID                                                             MLB
ITE_ITEM_DOM_DOMAIN_ID                                       MLB-CELLPHONES
ITE_ITEM_ID                                                      1722303517
ITE_ITEM_TITLE            celular samsung galaxy note20 tela 6 7 ram 8gb...
Name: 49, dtype: object

# 6. Cálculo completo

Pre procesamiento de títulos para agregar brand a títulos que no la tengan.

In [23]:
def title_brand(title, brands):
    for item in brands.split(' '):
        brand_add = ''
        if item not in title:
            brand_add += item
            
    return title + ' ' + brand_add

In [20]:
domains_list = list(domain_model_df.CAT_DOMAIN_ID.drop_duplicates())

chunk_size_whole = 25
chunks = int(len(domains_list) / chunk_size_whole)

print(chunks)

80


In [21]:
domains_list[25]

'MLB-SEX_BOARD_GAMES'

In [26]:
columns_names = ['id','category','subcategory','title','CAT_DOMAIN_ID','SIT_SITE_ID','ITE_ITEM_ID','ITE_ITEM_TITLE']

# Se hace el proceso iterativo por dominio (de los encontrados en los pasos anteriores)
# Arreglar el range
for segment in range(1,2):
    matches_df = pd.DataFrame(columns=columns_names)

    for cluster_number in range(segment*chunk_size_whole,(segment+1)*chunk_size_whole):
        
        # Lectura de títulos por dominio (Se trae el dominio completo)
        print(f'{cluster_number}. Executing neighbours search for domain {domains_list[cluster_number]}')
        query = """
            SELECT 
                SIT_SITE_ID,
                ITE_ITEM_DOM_DOMAIN_ID,
                ITE_ITEM_ID,
                ITE_ITEM_TITLE,
                
                BRAND_NAME,

            FROM
              `meli-bi-data.ML_TBL.ITEMS_SEGMENTOS`

            WHERE SIT_SITE_ID = 'MLB'
            AND GRUPO IN ('validos', 'incompletos')
            AND ITE_ITEM_DOM_DOMAIN_ID = '{dominios}'
        """.format(dominios = domains_list[cluster_number])

        meli_items_df = pd.DataFrame(bq_connec.execute_response(query))

        # Pre procesamiento de títulos por cluster y cálculo de embeddings
        if meli_items_df.shape[0] != 0:
            meli_items_df['BRAND_NAME'] = meli_items_df['BRAND_NAME'].apply(lambda x: '' if x is None else x)
            meli_items_df['BRAND_NAME'] = title_preprocesing(meli_items_df['BRAND_NAME'])
            meli_items_df['ITE_ITEM_TITLE'] = title_preprocesing(meli_items_df['ITE_ITEM_TITLE'])

            meli_items_df['ITE_ITEM_TITLE'] = meli_items_df.apply(lambda x: title_brand(x.ITE_ITEM_TITLE, x.BRAND_NAME), axis = 1)

            meli_items_df = meli_items_df.reset_index(drop = True)
            
            chunk_size = 10000
            chunks = int(np.ceil(meli_items_df.shape[0] / chunk_size))

            emb_meli_array = np.zeros((1,512))

            for i in range(chunks):
                t0 = time.time()
                emb_meli_array = np.append(
                    emb_meli_array, 
                    model.transform(meli_items_df.ITE_ITEM_TITLE.iloc[i*chunk_size:(i+1)*chunk_size].tolist()),
                    axis = 0
                )
                t1 = time.time()


            emb_meli_array = np.delete(emb_meli_array, 0, 0)

            # Cálculo del árbol annoy para vecinos cercanos en base a los embeddings de todo el dominio
            n, p = emb_meli_array.shape
            n_trees_from_data = int(np.sqrt(p))
            n_trees_from_data

            y = annoy.AnnoyIndex(p, "euclidean")
            for idx, row in enumerate(emb_meli_array):
                y.add_item(idx, row)

            y.build(n_trees_from_data)

            # Búsqueda de similaridad para cado uno de los items Nielsen
            search_df = domain_model_df[domain_model_df.CAT_DOMAIN_ID == domains_list[cluster_number]]

            for index, row in search_df.iterrows():

                idx, dist = y.get_nns_by_vector(emb_array[row.id], n=25, include_distances=True)

                neighbours_df = pd.DataFrame(columns=columns_names)

                for i, element in enumerate(dist):
                    if element < 0.6:            
                        aux_df = pd.DataFrame(pd.concat(
                            [
                                row[['id','category','subcategory','title','CAT_DOMAIN_ID']],
                                meli_items_df.iloc[idx[i]][['SIT_SITE_ID','ITE_ITEM_ID','ITE_ITEM_TITLE']]
                            ]
                        )
                        ).T

                        neighbours_df = pd.concat([aux_df[columns_names], neighbours_df], ignore_index=True)

                matches_df = pd.concat([neighbours_df, matches_df], ignore_index=True)


    # Se guardan los registros cada X dominios
    table = 'meli-bi-data.ML_DEV.TMP_ASSORTMENT_MLB_NN_MATCHES'

    schema = [bigquery.SchemaField("id", bigquery.enums.SqlTypeNames.INT64),
              bigquery.SchemaField("category", bigquery.enums.SqlTypeNames.STRING),
              bigquery.SchemaField("subcategory", bigquery.enums.SqlTypeNames.STRING),
              bigquery.SchemaField("title", bigquery.enums.SqlTypeNames.STRING),
              bigquery.SchemaField("CAT_DOMAIN_ID", bigquery.enums.SqlTypeNames.STRING),
              bigquery.SchemaField("SIT_SITE_ID", bigquery.enums.SqlTypeNames.STRING),
              bigquery.SchemaField("ITE_ITEM_ID", bigquery.enums.SqlTypeNames.INT64),
              bigquery.SchemaField("ITE_ITEM_TITLE", bigquery.enums.SqlTypeNames.STRING),
             ]

    for df_chunk in np.array_split(matches_df, 10):
        job_config = bigquery.LoadJobConfig(schema=schema)
        job_config.write_disposition = bigquery.WriteDisposition.WRITE_APPEND
        # Load data to BQ
        job = bq_connec.client.load_table_from_dataframe(df_chunk, table, job_config=job_config)
        job.result()
        
        
    print(f'chunk {segment} complete')

25. Executing neighbours search for domain MLB-SEX_BOARD_GAMES
26. Executing neighbours search for domain MLB-ARCADE_STICK_CASES
27. Executing neighbours search for domain MLB-FOOSBALL_TABLES
28. Executing neighbours search for domain MLB-ARCADE_PARTS_KITS
29. Executing neighbours search for domain MLB-BOARD_GAMES
30. Executing neighbours search for domain MLB-TOYS_AND_GAMES
31. Executing neighbours search for domain MLB-ROULETTE_SETS
32. Executing neighbours search for domain MLB-MULTIGAME_MACHINE_CARDS
33. Executing neighbours search for domain MLB-GAME_CONSOLES
34. Executing neighbours search for domain MLB-VIDEO_PLAYER_AND_GAME_CONSOLE_STANDS
35. Executing neighbours search for domain MLB-VIDEO_GAMES
36. Executing neighbours search for domain MLB-GAME_CONSOLE_CASES_AND_BAGS
37. Executing neighbours search for domain MLB-GAME_CONSOLES_VIDEO_GAMES_AND_ARCADE_MACHINES
38. Executing neighbours search for domain MLB-VIDEO_PLAYER_REMOTE_CONTROLS
39. Executing neighbours search for domain

In [35]:
matches_df.head(10)

id          category subcategory  \
0  4333  joias e relogios    pulseira   
1  4338  joias e relogios    pulseira   
2  4338  joias e relogios    pulseira   
3  4338  joias e relogios    pulseira   
4  4338  joias e relogios    pulseira   
5  4376  joias e relogios     brincos   
6  4376  joias e relogios     brincos   
7  4376  joias e relogios     brincos   
8  4376  joias e relogios     brincos   
9  4376  joias e relogios     brincos   

                                               title CAT_DOMAIN_ID  \
0  corrente meia cana 5mm + pulseira   banhado a ...     MLB-RINGS   
1  corrente cadeado 6mm + pulseira + pingente   b...     MLB-RINGS   
2  corrente cadeado 6mm + pulseira + pingente   b...     MLB-RINGS   
3  corrente cadeado 6mm + pulseira + pingente   b...     MLB-RINGS   
4  corrente cadeado 6mm + pulseira + pingente   b...     MLB-RINGS   
5  brinco de ouro 18k argola zirconia rodinado 0 ...     MLB-RINGS   
6  brinco de ouro 18k argola zirconia rodinado 0 ...     MLB-RINGS   
7  brinco de ouro 18k argola zirconia rodinado 0 ...     MLB-RINGS   
8  brinco de ouro 18k argola zirconia rodinado 0 ...     MLB-RINGS   
9  brinco de ouro 18k argola zirconia rodinado 0 ...     MLB-RINGS   

  SIT_SITE_ID ITE_ITEM_ID                                     ITE_ITEM_TITLE  
0         MLB  2184163141  pulseira 13mm grumet + anel letra banhado a ou...  
1         MLB  1947105161  kit cordao pulseira pingente e anel 8mm cadead...  
2         MLB  1419251958  kit cordao pulseira pingente e anel 8mm cadead...  
3         MLB  2064929721  kit cordao pulseira pingente e anel 8mm cadead...  
4         MLB  1886353199  kit cordao pulseira pingente e anel 8mm cadead...  
5         MLB  1883014677  anel solitario 5mm zirconia ouro 18k 20 2 06 g...  
6         MLB  2049204898  anel solitario coracao 8x8 mm em zirconia ouro...  
7         MLB  1186276040  anel solitario coracao 8x8 mm em zirconia ouro...  
8         MLB   957958717  anel 18k ouro branco goldfiled zirconia maior ...  
9         MLB  1110908752  anel solitario zirconia de 8mm banhado folhead...